project/yui0303/README.md

```diff
@@ -0,0 +1,225 @@
+1. Solve the characteristic polynomial.
+   - Maybe use QR Decomposition to find the eigenvalues
+2. Solve the Kernel space.
+   - Using Gaussian Elimination
+   - Other Methods?
+
+
+## How to diagonalize the matrix
+
+On $A_{n\times n}$ ,  
+If the eignvalues of $A$ is $\lambda_{1}, \lambda_{2}, ..., \lambda_{n}$ and the corresponding eigenvectors are $v_1, v_2, ..., v_n$, then the diagonalization of $A$ is shown as below.  
+
+$$
+P^{-1}AP=
+\begin{bmatrix}
+\lambda_{1} & & & & \\
+ & \lambda_{2} & & &\\
+ & & . & & \\
+ & & & . & \\
+ & & & & \lambda_{n}
+\end{bmatrix} 
+,
+P=
+\begin{bmatrix}
+  v_1 & v_2 & . & . & v_n
+\end{bmatrix}
+$$
+
+**prerequisite: A has n independent eigenvectors.**
+
+## Prospective Users
+
+The diagonal matrix has the property that all of its off-diagonal elements are zero. There are some prospective usages for related users are shown below.
+
+1. Solving Linear System
+   - Powers of matrix, i.e., Calculate $A^{100}$.
+   - SVD decomposition
+   - Markov Chains
+2. Some formula that contains eigenvalues or eigenvectors
+
+## System Architecture
+
+The flowchart of the project is shown in the image below.  
+
+![](./src/SystemArchitecture.jpg)
+
+The above algorithm will be encapsulated by c++, and the algorithm also could be called by python with pybind11. 
+
+The user interface may base on [Vue](https://vuejs.org/) or [React](https://zh-hant.legacy.reactjs.org/).
+
+### Constraints
+
+
+## API Description
+The estimated API is below:
+### c++
+```c++
+class Diagonalization{
+    public:
+        Diagonalization() = default;
+        Diagonalization(vector<vector<double>> input);
+        ~Diagonalization() = default;
+        void CalEigenvalues();
+        void CalEigenvectors();
+        bool Diagonalizable();
+        const vector<double> getEigenvalues const();
+        const vector<double> getEigenvectors const();
+        
+        // other member functions. To be continue...
+    private:
+        vector<vector<double>> arr;
+        vector<double> eignevalues;
+        vector<double> eignevectors;
+};
+
+```
+
+### python 
+The python API will call c++ API by pybind11, and the python API can accept the numpy format as input.
+
+## Engineering Infrastructure
+
+1. Documentation: README.md in Github Repository
+2. Automatic build system: Makefile or cmake
+3. Version control: Git
+4. Testing framework: Pytest
+5. Wrapping: Pybind11
+
+## Schedule
+
+- Week 1 (10/31):  
+  - Study domain knwoledges and choose the method of calculating eigenvalues.  
+  - Start to coding the function of `CalEigenvalues()`
+- Week 2 (11/7): 
+  - Finish the the function of `CalEigenvalues()`
+- Week 3 (11/14): 
+  - Start to coding the function of `CalEigenvectors()`
+  - Try how to partition the calcuation of eignevalues and eignevectors with multi-threading or other accerlation methods.
+- Week 4 (11/21): 
+  - Keep tring the accerlation of matrix operation.
+- Week 5 (11/28): 
+  - Binding the c++ API with pybind11
```

Comment:
It's not easy to write tests before you have the binding ready.  Testing speeds up your development.  Please think about how to make yourself more productive.